fix(node-packages): pin peer dependencies with undefined version

cgatt opened the pull request:
fixes #1611 

Peer dependencies added without a fixed version are being added to devDependencies in package.json as `name@^version` even with `pinnedDevDependency=true`, which is throwing a JSII warning. This change adds another pass through the peer dependencies in the post-synth step to pin those peer dependencies who's version is not known during presynth. While this causes a degree of code duplication, I decided against replacing the pre-synth peerDependency pass as it minimises the impact of this change on existing workflows.

This change is tested both in mocked unit tests and through yarn link to a local project, where it displayed the expected behaviour, including correctly respecting any existing versioning in `package.json`.

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Chriscbr commented on test/javascript/node-package.test.ts:
```diff
@@ -151,3 +151,156 @@ test("no install if package.json did not change at all", () => {
+
+  const pkgFile = readJsonSync(join(project.outdir, "package.json"));
+
+  expect(pkgFile.peerDependencies).toStrictEqual({ ms: "^2.1.3" });
+  expect(pkgFile.devDependencies).toStrictEqual({ ms: "2.1.3" });
```
Let's change this to some other random version number to make it clear that the mock is working and the test isn't dependent on the actual latest version of "ms"

cgatt replied:
Done

Chriscbr commented on test/javascript/node-package.test.ts:
```diff
@@ -151,3 +151,156 @@ test("no install if package.json did not change at all", () => {
 });
+
+test('"*" peer dependencies are pinned in devDependencies', () => {
+  jest.spyOn(util, "exec").mockImplementation((command, options) => {
+    expect(command.startsWith("yarn install")).toBeTruthy();
```
Add a comment explaining why this mock is needed

Chriscbr commented on test/javascript/node-package.test.ts:
```diff
@@ -151,3 +151,156 @@ test("no install if package.json did not change at all", () => {
+    throw new Error(`unexpected version: ${ver}`);
+  });
+
+  const project = new Project({ name: "test" });
+  const pkg = new NodePackage(project);
+
+  const orig = {
+    name: "test",
+    devDependencies: {
+      ms: "^2.2.2",
+    },
+    main: "lib/index.js",
+    license: "Apache-2.0",
+    version: "0.0.0",
+    "//": '~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".',
+  };
```
I'm not sure if I understand this test case. It looks like we are asserting that if the user happens to have "ms" as a dev dependency (but not configured by projen), that projen will not change that after synthesizing? I'm not sure if that is an appropriate promise to make in our API -- either the user should be managing all of their dependencies through projen, or none of them.

cgatt replied:
This was included to mimic the current behaviour, on line 979 of node-package.ts:
`// respect user version if it's not '*' or if current version is undefined`
I'd be more than happy for the behaviour of the peer pinning to behave differently in this case if that's preferable?

Chriscbr commented on test/javascript/node-package.test.ts:
```diff
@@ -151,3 +151,156 @@ test("no install if package.json did not change at all", () => {
+    throw new Error(`unexpected version: ${ver}`);
+  });
+
+  const project = new Project({ name: "test" });
+  const pkg = new NodePackage(project);
```
If these test cases are assuming that pinnedDevDependencies is true, we should specify that explicitly in the tests when initializing NodePackage, and not take the default for granted. If the test case assertions here are irrespective of the setting of pinnedDevDependencies, then I think it might make sense to show that explicitly (e.g. by dividing this into one test where pinnedDevDependencies is true, and one test where it's false).

cgatt replied:
Done

Chriscbr reviewed (changes requested):
Could you create a bug report first (and then link this PR to it)? It's really helpful for us to have an explanation of the issue by itself -- something of the flavor of "when I did X I expected Y, but instead Z happened".

cgatt commented:
I've raised an issue and linked to it in the PR description, and modified the tests.
Went slightly overboard because the repeated code for mocking yarn install was annoying me, so I moved it into a utility function which has a bit of extra logic for handling multiple dependencies fields and installing a version that meets all given dependency ranges. Wasn't needed for this, but I saw it being a problem for future tests. This is why I've slightly changed an existing test (line 162), as I made the fake `yaml.lock` file a little closer to the real format.
* Tests have been modified to each use arbitrary dep versions to ensure no interaction/external calls.
* Tests explicitly set pinnedDevDependencies to true
* Additional test added for pinnedDevDependencies = false
* Additional comments included

Chriscbr commented on test/javascript/node-package.test.ts:
```diff
@@ -151,3 +215,125 @@ test("no install if package.json did not change at all", () => {
+  expect(pkgFile.devDependencies).toStrictEqual({ ms: "^1.3.0" });
+});
+
+test("devDependencies are not pinned by peerDependencies if dependency exists", () => {
```
```suggestion
test("devDependencies are not pinned by peerDependencies if a regular (runtime) dependency also exists", () => {
```

Chriscbr reviewed (changes requested):
Thanks for your patience @cgatt. I think is almost ready to get merged in, just have a few comments to improve readability in a few places.

Overall the code in projen for resolving dependencies is a bit confusing to grok / feels like it could benefit from some kind of refactoring, but that makes more sense for a separate PR. I appreciate you writing out several test cases and keeping the actual fix here pretty self-contained. 👍
